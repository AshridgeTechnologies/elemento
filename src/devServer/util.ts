import fs from 'fs'
import axios from 'axios'

export async function getModule(moduleUrl: string, moduleLocalPath: string, moduleName: string) {
    const moduleAlreadyDownloaded = fs.existsSync(moduleLocalPath)
    const moduleDate = moduleAlreadyDownloaded ? fs.statSync(moduleLocalPath).mtime : new Date(0)
    try {
        const resp = await axios.get(moduleUrl, {
            headers: {
                'If-Modified-Since': moduleDate.toUTCString(),
                'Cache-Control': 'max-age=31536000'
            },
            validateStatus: status => status >= 200 && status <= 304
        })
        if (resp.status === 304) {
            console.info(`Latest ${moduleName} already downloaded from`, moduleUrl)
            return true
        }
        if (resp.status !== 200) {
            throw new Error(resp.status + ' ' + resp.statusText)
        }
        const moduleContents = await resp.data
        await fs.promises.writeFile(moduleLocalPath, moduleContents)
        console.info(`Downloaded latest ${moduleName} from`, moduleUrl)
        return true
    } catch (e) {
        console.warn(`Cannot download latest ${moduleName} from`, moduleUrl, (e as Error).message)
        if (moduleAlreadyDownloaded) {
            console.warn(`Using previously downloaded version of ${moduleName}`)
            return true
        } else {
            return false
        }
    }
}