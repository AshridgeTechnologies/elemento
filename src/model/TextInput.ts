import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}