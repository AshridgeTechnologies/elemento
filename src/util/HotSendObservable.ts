import SendObservable, {Observer} from './SendObservable'

const _NOT_SET = 'Not set'
export default class HotSendObservable<T> extends SendObservable<T> {

    private latestValue: T | typeof _NOT_SET = _NOT_SET

    onSubscribe(observer: Observer<T>): () => void {
        if (this.latestValue !== _NOT_SET) {
            observer.next(this.latestValue)
        }

        return super.onSubscribe(observer)
    }

    send(value: T) {
        this.latestValue = value
        super.send(value)
    }

}

