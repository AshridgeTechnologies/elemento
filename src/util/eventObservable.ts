import Observable from 'zen-observable'
import {without} from 'ramda'
import {Observer} from './SendObservable'

type EventSource = {addEventListener: typeof Window.prototype.addEventListener, removeEventListener: typeof Window.prototype.removeEventListener}

export default function eventObservable(eventSource: EventSource, eventType: string, getDataFromEvent: (evt: Event) => any)  {
    let observers: Observer<Event>[] = []
    const handler = (event: Event) => {
        const data = getDataFromEvent(event)
        observers.forEach( obs => obs.next(data) )
    }

    return new Observable(observer => {

        if (observers.length === 0) {
            eventSource.addEventListener(eventType, handler)
        }
        observers.push(observer)

        // return unsubscribe function
        return () => {
            observers = without([observer], observers)
            if (observers.length === 0) {
                eventSource.removeEventListener(eventType, handler)
            }
        }
    })
}
