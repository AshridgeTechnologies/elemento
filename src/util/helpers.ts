export function definedPropertiesOf(obj: object) {
    return Object.fromEntries(Object.entries(obj).filter(([,val]) => val !== undefined))
}