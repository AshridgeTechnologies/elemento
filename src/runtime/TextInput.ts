import React from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../util/helpers'

type Properties = {label?: string, initialValue?: string, maxLength?: number}

export default function TextInput({initialValue = '', maxLength, label}: Properties) {
    const maxLengthProps = maxLength !== undefined ? {inputProps: {maxLength: maxLength}} : {}
    const optionalProps = definedPropertiesOf({label})
    return React.createElement(TextField, {
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value: initialValue,
        ...maxLengthProps,
        ...optionalProps
    })
}
