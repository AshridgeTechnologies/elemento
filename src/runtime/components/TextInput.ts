import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {updateState} from '../appData'
import {valueOfProps} from '../runtimeFunctions'

type Properties = {state: {value?: string, _path: string, _controlValue: string | null}, label?: string, maxLength?: number, width?: string | number, multiline?: boolean}

export default function TextInput({state, ...props}: Properties) {
    const {maxLength, label, multiline, width} = valueOfProps(props)
    const inputProps = maxLength !== undefined ? {inputProps: {maxLength}} : {}
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}
    const optionalProps = definedPropertiesOf({label, multiline})
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        updateState(path, {value: updateValue })
    }

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
}
