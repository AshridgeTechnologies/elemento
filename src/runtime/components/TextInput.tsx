import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {PropVal, valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {TextType} from '../../shared/types'
import {pick} from 'ramda'
import {InputWithInfo} from './InputWithInfo'


type Properties = {path: string, label?: PropVal<string>, width?: PropVal<string | number>, multiline?: PropVal<boolean>, readOnly?: PropVal<boolean>}

export default function TextInput({path, ...props}: Properties) {
    const {label, multiline: multilineProp, width, readOnly} = valueOfProps(props)
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}

    const state = useGetObjectState<TextInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required
    const multiline = dataType?.format === 'multiline' || multilineProp
    const multilineProps = multiline ? {minRows: 2, maxRows: 10} : {}
    const optionalProps = definedPropertiesOf({label, multiline, ...multilineProps})

    const dataTypeProps = definedPropertiesOf(pick(['maxLength', 'minLength'], dataType ?? {}))
    const format = dataType?.format
    if (format && ['url', 'email'].includes(format)) {
        dataTypeProps.type = format
    }
    const inputPropsValues = {...dataTypeProps, ...definedPropertiesOf({readOnly}), }
    const inputProps = Object.keys(inputPropsValues).length > 0 ? {inputProps: inputPropsValues} : {}

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const formControl =  React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
    return InputWithInfo({description: dataType?.description, required, width, formControl})
}

export class TextInputState extends InputComponentState<string, TextType> {
    defaultValue = ''
}

TextInput.State = TextInputState