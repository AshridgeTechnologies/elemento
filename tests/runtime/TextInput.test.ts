/**
 * @jest-environment jsdom
 */

import {createElement} from 'react'
import TextInput from '../../src/runtime/TextInput'
import {snapshot} from '../util/testHelpers'

test('TextInput element produces output with properties supplied',
    snapshot(createElement(TextInput, {initialValue: 'Hi there!', maxLength: 10, label: 'Item Description'}))
)

test('TextInput element produces output with default values where properties omitted',
    snapshot(createElement(TextInput, {}))
)