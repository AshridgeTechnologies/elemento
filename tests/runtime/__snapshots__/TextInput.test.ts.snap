// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1a1ihhj"
      >
        <span
          className="notranslate"
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`TextInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
  >
    Item Description
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      maxLength={10}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value="Hi there!"
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-186xcr5"
      >
        <span>
          Item Description
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
